UPSTREAM: <carry>: openshift: Implement scale from zero

elmiko opened the pull request:
This allows a Machine{Set,Deployment} to scale up/down from 0,
providing the following annotations are set:

```yaml
apiVersion: v1
items:
- apiVersion: machine.openshift.io/v1beta1
  kind: MachineSet
  metadata:
    annotations:
      machine.openshift.io/cluster-api-autoscaler-node-group-min-size: "0"
      machine.openshift.io/cluster-api-autoscaler-node-group-max-size: "6"
      machine.openshift.io/vCPU: "2"
      machine.openshift.io/memoryMb: 8G
      machine.openshift.io/GPU: "1"
      machine.openshift.io/maxPods: "100"
```

please note that `machine.openshift.io/maxPods` is not required for scale to/from zero operations.

this PR is a continuation of the work started in #110 

Counterpart PRs:
openshift/cluster-api-provider-aws#301
openshift/cluster-api-provider-azure#112
Design details:
openshift/enhancements#186

enxebre commented:
Counterpart PRs
openshift/cluster-api-provider-aws#301
openshift/cluster-api-provider-azure#112

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils.go:
```diff
@@ -20,12 +20,19 @@ import (
 const (
 	nodeGroupMinSizeAnnotationKey = "machine.openshift.io/cluster-api-autoscaler-node-group-min-size"
 	nodeGroupMaxSizeAnnotationKey = "machine.openshift.io/cluster-api-autoscaler-node-group-max-size"
+
```
can we update this annotations to match what's given here
openshift/cluster-api-provider-aws#301
openshift/cluster-api-provider-azure#112

elmiko replied:
similar question to the other annotations, should we make these exportable from the autoscaler so that the providers can import them?

elmiko replied:
i am seeing the comments from the other pr. no further clarification needed =)

JoelSpeed replied:
These `nodeGroupInstance...` annotations are being used to parse, but the `...Key` ones are what will exist, so I think the parsing needs updating and these 4 consts can be dropped (if I've understood correctly)

elmiko replied:
that makes sense. thanks!

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils.go:
```diff
@@ -143,3 +152,38 @@ func machineSetIsOwnedByMachineDeployment(machineSet *MachineSet, machineDeploym
 }
+
+func scaleFromZeroEnabled(annotations map[string]string) bool {
+	if val, exists := annotations[nodeGroupScaleFromZero]; exists {
```
can we get this implicitly by checking the existence  of the annotations?

elmiko replied:
makes sense, i have a couple questions:

1. must all these annotations be defined? (eg can a resource have cpu and mb but not gpu)
1. should i make these keys package level exports so that the providers can vendor them in? (my concern is these variables becoming out of sync between repos)

elmiko replied:
added a fix for this

frobware replied:
> can we get this implicitly by checking the existence of the annotations?

I was previously using this as a brute-force switch to test the behaviour.

However, what's unclear is does the presence of some/all/one annotation in the scale-from-zero-set imply that this machine set can scale from zero?

enxebre replied:
We have control over which providers we officially support autoscaling from/to zero as a product.
We signal this to the autoscaler by enabling a controller in any given platform which sets the well known annotations we choose. https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/actuators/machineset/controller.go#L108-L110
i.e 
```
if has cpuKey && memKey && gpuKey
 return true
```

maxPods should optional and default to 255 here, see last bullet point https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/autoscaling-from-to-zero.md#machine-api-autoscaler-provider

JoelSpeed replied:
I think GPU should potentially be optional. I've started working on the GCP implementation and as far as I can tell, at present MAPI doesn't support adding GPUs for GCP instances. Seems odd for the controller to just permanently set that value to zero?

elmiko replied:
gpu definitely _seems_ like it should be optional, i'll make some adjustments. thanks!

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -18,16 +18,23 @@ package openshiftmachineapi
 	machineAnnotationKey       = "machine.openshift.io/machine"
 	debugFormat                = "%s (min: %d, max: %d, replicas: %d)"
+
+	// This default for the maximum number of pods comes from the machine-config-operator
```
can we include a link here? e.g https://github.com/openshift/machine-config-operator/blob/2f1bd6d99131fa4471ed95543a51dec3d5922b2b/templates/worker/01-worker-kubelet/_base/files/kubelet.yaml#L19

elmiko replied:
absolutely, i actually debated dropping a link there lol

JoelSpeed commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils.go:
```diff
@@ -143,3 +151,42 @@ func machineSetIsOwnedByMachineDeployment(machineSet *MachineSet, machineDeploym
+	mem := annotations[memoryKey]
+	gpu := annotations[gpuKey]
+
+	if cpu != "" && mem != "" && gpu != "" {
```
Suggestion to only require CPU and Memory since GPU can have a sane default of `0` and we don't currently support GPUs across all providers (eg GCP)
```suggestion
	if cpu != "" && mem != ""  {
```

JoelSpeed commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils.go:
```diff
@@ -143,3 +151,42 @@ func machineSetIsOwnedByMachineDeployment(machineSet *MachineSet, machineDeploym
+	if val, exists := annotations[cpuKey]; exists && val != "" {
+		return resource.ParseQuantity(val)
+	}
+	return zeroQuantity.DeepCopy(), nil
```
Wonder if it might be worth extracting this to something like below and calling that from each of the methods listed here since the only difference between them is the key passed into the map to check?
```
func parseKey(annoations map[string]string, key string) (resource.Quantity, error) {
    if val, exists := annotations[key]; exists && val != "" {
	return resource.ParseQuantity(val)
    }
    return zeroQuantity.DeepCopy(), nil
}
```

elmiko replied:
makes good sense, i'll refactor it

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -212,9 +221,59 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+	pod, err := ng.scalableResource.InstanceMaxPodsCapacity()
+	if err != nil {
+		return nil, err
+	}
```
i _think_ we do

elmiko replied:
well, it certainly sounded reasonable when i said it earlier, but now that i am digging through the code i am not finding a gpu-related resource key in the corev1 stuff.

is there a document that describes these capacity values?

elmiko replied:
perhaps this is they key i want?

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L365

JoelSpeed replied:
That does seem to be the only key I can find, all the other providers use it so I guess it makes sense to use it

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -265,3 +324,26 @@ func newNodegroupFromMachineDeployment(controller *machineController, machineDep
+	m := make(map[string]string)
+	m[kubeletapis.LabelArch] = cloudprovider.DefaultArch
+	m[kubeletapis.LabelOS] = cloudprovider.DefaultOS
+	//m[kubeletapis.Hostname] = nodeName
```
not sure, maybe @frobware has a quick answer

frobware replied:
Do what AWS does: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L397. I may have been experimenting with this to see if it was really needed.

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -265,3 +324,26 @@ func newNodegroupFromMachineDeployment(controller *machineController, machineDep
 	}, nil
 }
+
+func joinStringMaps(labels ...map[string]string) map[string]string {
```
good question, i'll dig in a little

frobware replied:
> What's the benefit of using this over `cloudprovider.JoinStringMaps`? Looks like it's only being used in one place on L#271 and it's only joining two maps so I'm not sure it's any more efficient?

Look to the AWS provider; the patterns for supporting TemplateNodeInfo was taken from there, with some adjustments/leftovers as I discovered how this should be implemented.

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L377

cc @elmiko

elmiko replied:
after looking at this a bunch, i think we can just use the `cloudprovider.JoinStringMaps`

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -212,9 +221,59 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+
+	if pod.IsZero() {
+		pod = *resource.NewQuantity(defaultMaxPods, resource.DecimalSI)
+	}
```
makes sense to me

frobware commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup_test.go:
```diff
@@ -168,8 +168,10 @@ func TestNodeGroupNewNodeGroupConstructor(t *testing.T) {
 
-		if _, err := ng.TemplateNodeInfo(); err != cloudprovider.ErrNotImplemented {
-			t.Error("expected error")
+		if ng.scalableResource.CanScaleFromZero() {
```
This is/was my "big switch" whilst experimenting. It can go.

frobware commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go:
```diff
@@ -212,9 +221,59 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+	node := corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:     nodeName,
+			SelfLink: fmt.Sprintf("/api/v1/nodes/%s", nodeName),
```
Again, how this function should work is all taken from the AWS provider:

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go

frobware replied:
Correct link/ref: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L354

JoelSpeed replied:
Oh strange, they're setting the field all over this codebase, though as far as I can tell, it's never read anywhere. If we wan't to keep it looking like AWS then it's fine to leave, doesn't cause any harm

frobware replied:
Having said "do what AWS do", it's also reasonable to question that. It's worth experimenting to see what happens if you don't set this.

elmiko replied:
given this line from the current kube api docs, i feel ok dropping `selfLink`

> DEPRECATED Kubernetes will stop propagating this field in 1.20 release and the field is planned to be removed in 1.21 release.

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_machinedeployment.go:
```diff
@@ -134,3 +137,31 @@ func newMachineDeploymentScalableResource(controller *machineController, machine
+}
+
+func (r machineDeploymentScalableResource) Labels() map[string]string {
+	return cloudprovider.JoinStringMaps(r.machineDeployment.Labels, r.machineDeployment.Spec.Template.Labels, r.machineDeployment.Spec.Template.Spec.Labels)
```
i'm not quite sure about this one, still researching

enxebre replied:
The ones propagated on node creation are `machineDeployment.Spec.Template.Labels`

https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/machine/v1beta1/machine_types.go#L72-L81

elmiko replied:
not sure i fully understand this from the code linked, but i've removed the join and changed to the return to `machineDeployment.Spec.Template.Spec.Labels`

elmiko replied:
see https://github.com/openshift/kubernetes-autoscaler/pull/137#discussion_r398171772

elmiko commented:
updated to incorporate the comments here, and squashed my additions into a single commit.

elmiko commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils_test.go:
```diff
@@ -369,3 +370,197 @@ func TestUtilMachineSetMachineDeploymentOwnerRef(t *testing.T) {
+		},
+		enabled: false,
+	}, {
+		description: "matching key, value=true annotation",
```
agreed, i will change it

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go:
```diff
@@ -443,7 +443,9 @@ func (c *machineController) machineSetNodeGroups() ([]*nodegroup, error) {
-		if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
+		if ng.scalableResource.CanScaleFromZero() {
+			nodegroups = append(nodegroups, ng)
+		} else if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
```
shouldn't the check `ng.MaxSize()-ng.MinSize() > 0` be at the top?
So no matter this node group can scale from zero or not. If ``ng.MaxSize()-ng.MinSize() > 0` is false it should not be included.

enxebre replied:
i.e.
```
if ng.MaxSize()-ng.MinSize() > 0
    if ng.scalableResource.CanScaleFromZero()
             append
    else if pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0
             append
```

enxebre replied:
The same logic we settle here needs to be included for `machineDeploymentNodeGroups`

elmiko replied:
this logic makes sense to me, if essentially if MaxSize < MinSize then we have bigger problems.

i like your solution of gating on MaxSize - MinSize > 0 first, i will change this code to follow that.

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go:
```diff
@@ -540,12 +541,11 @@ func (c *machineController) nodeGroupForNode(node *corev1.Node) (*nodegroup, err
-	if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
-		return nil, nil
+	if !nodegroup.scalableResource.CanScaleFromZero() {
+		if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
```
shouldn't this be always the cases regardless of `CanScaleFromZero`? max-min must be at least 1.

enxebre replied:
https://github.com/openshift/kubernetes-autoscaler/pull/137/files#r394255920

elmiko replied:
your reasoning makes sense to me here, i'm going to drop the check against CanScaleFromZero and just rely on the MaxSize-MinSize > 1.

elmiko replied:
one question though, should we check for the situation where `MaxSize-MinSize < 1 && MaxSize != 0` ?

essentially, do we need to return the nodegroup when max and min are equal, but not zero?

_i'm guessing not, since we won't be autoscaling in those cases, but i want to make sure_

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go:
```diff
@@ -525,11 +527,10 @@ func (c *machineController) nodeGroupForNode(node *corev1.Node) (*nodegroup, err
-			// least 1.
-			if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
-				return nil, nil
+			if !nodegroup.scalableResource.CanScaleFromZero() {
```
It seems to me the comment is misleading?
shouldn't this be always the cases regardless of CanScaleFromZero? max-min must be at least 1.
cc @frobware

frobware replied:
What happens, even today, if you set min:0, max:0? we get 0, but prior to this effort we had to have at least 1 node.

enxebre replied:
mm I'm struggling to see how this relates to this PR/support for scaling from zero.
Before this PR you could  have `min:1, max:1` which would fail the check. Why would we change that for the case where `min:0`?
Seems to me `if nodegroup.MaxSize()-nodegroup.MinSize() < 1` is just preventing you from having the same max and min.

elmiko replied:
i've changed this to be similar to the logic in the machinedeployments section, so basically only checking if MaxSize-MinSize < 1. hopefully i've understood the logic properly XD

enxebre commented on cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_machineset.go:
```diff
@@ -119,3 +122,31 @@ func newMachineSetScalableResource(controller *machineController, machineSet *Ma
+
+func (r machineSetScalableResource) Labels() map[string]string {
+	return cloudprovider.JoinStringMaps(r.machineSet.Labels, r.machineSet.Spec.Template.Labels, r.machineSet.Spec.Template.Spec.Labels)
+}
```
The only ones propagated on node creation are `machineDeployment.Spec.Template.Spec.Labels` so no need to join.

https://github.com/openshift/machine-api-operator/blob/master/pkg/apis/machine/v1beta1/machine_types.go#L72-L81

elmiko replied:
so, it turns out that reducing this to just `r.machineSet.Spec.Template.Spec.Labels` causes our test to not work. i'm not sure which is more correct though, keep the label join here or fix the test.

enxebre commented:
This is looking great @elmiko. We need to rebase as https://github.com/openshift/kubernetes-autoscaler/pull/137 got in.
In parallel can we put a PR against https://github.com/openshift/cluster-api-actuator-pkg/blob/master/pkg/autoscaler/autoscaler.go#L209 to include and validate scaling from zero scenario?

elmiko commented:
rebased, still working on the outstanding comments

enxebre commented:
can we conflate both commits into one? the split is meaningless since we never had support for the first one.

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -103,8 +112,10 @@ func (ng *nodegroup) DeleteNodes(nodes []*corev1.Node) error {
 
-	if replicas-int32(len(nodes)) <= 0 {
-		return fmt.Errorf("unable to delete %d machines in %q, machine replicas are <= 0 ", len(nodes), ng.Id())
+	if !ng.scalableResource.CanScaleFromZero() {
```
`if ng.scalableResource.CanScaleFromZero()` is true we still  want to validate and error `if replicas-int32(len(nodes)) < 0 {`

enxebre replied:
https://github.com/openshift/kubernetes-autoscaler/pull/137/files#r396385396

elmiko replied:
seems like we need to check both conditions depending on if it can scale or not. eg:

```
if CanScaleFromZero and replicas-len(nodes) < 0
then
  error, deleting too many
else if !CanScaleFromZero and replicas-len(nodes) <= 0
then
  error, deleting too many
```

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -566,12 +567,11 @@ func (c *machineController) nodeGroupForNode(node *corev1.Node) (*nodegroup, err
-	// that has a scale size of at least 1.
-	if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
-		return nil, nil
+	if !nodegroup.scalableResource.CanScaleFromZero() {
```
why do we only check this  `if !nodegroup.scalableResource.CanScaleFromZero()`?

enxebre replied:
https://github.com/openshift/kubernetes-autoscaler/pull/137/files#r396385396

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -471,7 +471,9 @@ func (c *machineController) machineSetNodeGroups() ([]*nodegroup, error) {
-		if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
+		if ng.scalableResource.CanScaleFromZero() {
+			nodegroups = append(nodegroups, ng)
+		} else if ng.MaxSize()-ng.MinSize() > 0 && pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
```
we still want to check `ng.MaxSize()-ng.MinSize() > 0` when CanScaleFromZero,  we don't want to add groups where min=max. This will effectively add every machineSet
```
I0323 12:33:09.401145   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1e (min: 0, max: 0, replicas: 0)
I0323 12:33:09.401154   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1f (min: 0, max: 0, replicas: 0)
I0323 12:33:09.401161   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1a (min: 0, max: 0, replicas: 1)
I0323 12:33:09.401168   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1b (min: 0, max: 0, replicas: 1)
I0323 12:33:09.401175   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1c (min: 0, max: 0, replicas: 1)
I0323 12:33:09.401182   21597 clusterapi_provider.go:71] discovered node group: openshift-machine-api/agl230320-cdzht-worker-us-east-1d (min: 0, max: 0, replicas: 0)
```


```
ng.MaxSize()-ng.MinSize() > 0`

enxebre replied:
i.e we should:
```
		if ng.MaxSize()-ng.MinSize() > 0 {
			if ng.scalableResource.CanScaleFromZero() {
				nodegroups = append(nodegroups, ng)
			} else if pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0) > 0 {
				nodegroups = append(nodegroups, ng)
			}
		}
```

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -551,11 +553,10 @@ func (c *machineController) nodeGroupForNode(node *corev1.Node) (*nodegroup, err
-			// least 1.
-			if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
-				return nil, nil
+			if !nodegroup.scalableResource.CanScaleFromZero() {
```
same that https://github.com/openshift/kubernetes-autoscaler/pull/137/files#r396385396

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go:
```diff
@@ -153,3 +161,36 @@ func normalizedProviderString(s string) normalizedProviderID {
+}
+
+func parseMemoryCapacity(annotations map[string]string) (resource.Quantity, error) {
+	return parseKey(annotations, memoryKey)
```
this is comparing not matching formats, i.e this would expect 
```
return resource.ParseQuantity(fmt.Sprintf("%vMi", val))
```

elmiko replied:
do we need to handle the case where a user puts the unit type in their annotation?

enxebre replied:
The user does not set this value, our controllers do. We set a contract that we pass a number representing Mb `machine.openshift.io/memoryMb` any additional check here wouldn't hurt though.

elmiko commented:
> This is looking great @elmiko. We need to rebase as #137 got in.

sounds good, i will rebase again and squash the 2 commits into one.

> In parallel can we put a PR against https://github.com/openshift/cluster-api-actuator-pkg/blob/master/pkg/autoscaler/autoscaler.go#L209 to include and validate scaling from zero scenario?

i am still working through my understanding of the test framework, but i will give it my best =)

elmiko commented:
rebased and squashed, also addressed the comments. still have a question about parsing the memory key.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters.go:
```diff
@@ -42,6 +42,22 @@ func newMachineDeploymentFromUnstructured(u *unstructured.Unstructured) *Machine
+	if err == nil && found {
+		machineSpecUnstructured := unstructured.Unstructured{Object: machineSpec}
+		machineDeployment.Spec.Template.Spec.Labels = machineSpecUnstructured.GetLabels()
+		taints, _, _ := unstructured.NestedSlice(machineSpec, "Taints")
```
When using the unstructured, everything is the way it would be written in `json`, so I think this should be lower cased
```suggestion
		taints, _, _ := unstructured.NestedSlice(machineSpec, "taints")
```

elmiko replied:
yeah, i'll go back and clean this up. i wanted to share my early work to see if my approach was proper.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters.go:
```diff
@@ -69,6 +85,22 @@ func newMachineSetFromUnstructured(u *unstructured.Unstructured) *MachineSet {
+	if err == nil && found {
+		machineSpecUnstructured := unstructured.Unstructured{Object: machineSpec}
+		machineSet.Spec.Template.Spec.Labels = machineSpecUnstructured.GetLabels()
+		taints, _, _ := unstructured.NestedSlice(machineSpec, "Taints")
```
Same comment as above
```suggestion
		taints, _, _ := unstructured.NestedSlice(machineSpec, "taints")
```

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters.go:
```diff
@@ -42,6 +42,22 @@ func newMachineDeploymentFromUnstructured(u *unstructured.Unstructured) *Machine
 		Status: MachineDeploymentStatus{},
 	}
 
+	machineSpec, found, err := unstructured.NestedMap(u.Object, "spec", "template", "spec")
```
Just thinking, there's an awful lot of duplication in this file, might be good to refactor at some point so we have things like `newMachineSpecFromUnstructured` that can be reused within the machine, machineset and machinedeployment converters 🤔

elmiko replied:
++, i will definitely clean this up

elmiko replied:
i cleaned it up some, let me know if this it's looking better.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters_test.go:
```diff
@@ -0,0 +1,84 @@
+		t.Error("MachineSet.Namespace improperly set")
+	}
+
+	if machineSet.Spec.Template.Spec.Labels["foo"] != "bar" {
```
Should add a test for taints too? 

Also, might be easier to just create an expected object and compare the two, since we should be able to construct a full object `&Machine{...}` with the expected outcome, then adding future checks would be a case of updating the input and expected output `Machine`, though it would make it harder to compare when there's a failure, not sure 🤔

elmiko replied:
i am still working on the rest of the test suite, i just wanted to get some opinions on direction before going further.

as for comparing the objects, i thought about that but as i was still building the test i just took the slow approach you see here.

elmiko replied:
i added tests for taints, but the Time values are driving me slightly mad currently

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters_test.go:
```diff
@@ -0,0 +1,84 @@
+				},
+			},
+		},
+	}
```
Does this cover all of the fields in a `MachineSet`?

elmiko replied:
not yet

elmiko replied:
so, i have some questions about how deep we should go with these things. as i descend into more and more objects i start to wonder if we should be testing the basic copy stuff?

i am more feeling that we should test the extra bits we have added, eg taint replications and label deep copy, rather than testing to see if the all the fields are just copied. that said, i have left the extra tests i added because it is probably better for coverage to have this stuff.

elmiko replied:
as per our discussions, i am adding more tests here and making them more-or-less a bunch of brute force tests. i have also added a jira ticket (see OCPCLOUD-826) to address improving the tests.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_converters_test.go:
```diff
@@ -0,0 +1,84 @@
+	namespace  = "Bar"
+)
+
+func TestConverterNewMachineSetFromUnstructured(t *testing.T) {
```
Will there be tests for the other converters? Could maybe follow up with those 🤔 

One interesting test to add on top of this as well may be to create a complete {Machine, MachineSet, MachineDeployment} and then convert it to unstructured and back again using the converters in this package and check the conversion doesn't lose any data?

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_machineset.go:
```diff
@@ -120,3 +122,31 @@ func newMachineSetScalableResource(controller *machineController, machineSet *Ma
+
+func (r machineSetScalableResource) Taints() []apiv1.Taint {
+	return r.machineSet.Spec.Template.Spec.Taints
+}
```
Nit: These are the opposite way around from in the machinedeployment, very minor but might make updates easier if they diff nicely 😛

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -212,9 +223,68 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+		gpuapis.ResourceNvidiaGPU: gpu,
+	}
+
+	nodeName := fmt.Sprintf("%s-asg-%d", ng.Name(), rand.Int63())
```
Where does `asg` come from in this? 🤔 Sounds like an AWS term to me 😅

elmiko replied:
good question, i'll work on find a doc link or something

elmiko replied:
ok, dug in a little bit.

i have a feeling that "asg" is inspired by [AutoScalingGroups](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html) from Amazon.

and indeed, this is how it is implemented in the aws provider:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L350

but, and probably due to copying by other vendors this also appears in:
AliCloud:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/alicloud/alicloud_manager.go#L212

Azure:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/azure/azure_scale_set.go#L517

BaiduCloud:
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/baiducloud/baiducloud_manager.go#L244

i suspect all these implementations are inspired by the aws provider, as was ours.

GCE for example uses `-template-`,
https://github.com/openshift/kubernetes-autoscaler/blob/master/cluster-autoscaler/cloudprovider/gce/templates.go#L126

i'm not sure if there is a "correct" answer here, but i am leaning towards leaving this the way it is unless we have a clear idea of what we would prefer. i am open to suggestions.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -265,3 +335,11 @@ func newNodegroupFromMachineDeployment(controller *machineController, machineDep
 	}, nil
 }
+
+func buildGenericLabels(nodeName string) map[string]string {
```
Might be worth a comment to explain what these labels are and why we are adding them? Like, why did we choose these three in particular to add?

elmiko replied:
so, i looked around a little bit and i see these values used by some of the other cloud providers for the autoscaler, but i don't see them used in all of the cluster-api provider implementations.

at a speculation, @frobware might have copied these as they are present in the other providers but they are not directly used by the cluster-api implementation.

i will add a TODO here, i think we should probably re-assess their inclusion after further testing.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils.go:
```diff
@@ -153,3 +162,46 @@ func normalizedProviderString(s string) normalizedProviderID {
+	if exists && val != "" {
+		if _, err := strconv.Atoi(val); err == nil {
+			// value is a number, we will append "Mi" as the unit type
+			val = fmt.Sprintf("%sMi", val)
```
I guess this is because the cloud provider controllers aren't labelling them correctly? To me, this feels like we should fix that in the labelling rather than here, WDYT @elmiko @enxebre ?

My consideration is we should try to keep the format the same as expected by Kubernetes itself (https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory) and shouldn't be expecting it to be a number of megabytes

elmiko replied:
++ agreed about the provider using values that follow the kubernetes guide.
currently the aws provider is just using a number, i'll take a look at that code to see if we could improve it.

enxebre replied:
Yes, I'd like to get this PR merged in a way which support compatibility i.e as it's now (checks if the value has a format and append otherwise).
Then merge the e2e PR.
Then do the changes on the actuators to set M/Mi driven by e2e running

JoelSpeed replied:
Cool, so long as there is a plan and we don't lose track, happy for this change to be a follow up

elmiko replied:
i will add a TODO note here

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils_test.go:
```diff
@@ -400,3 +401,202 @@ func TestUtilNormalizedProviderID(t *testing.T) {
+		annotations:      map[string]string{cpuKey: "123m"},
+		expectedError:    false,
+		expectedQuantity: resource.MustParse("123m"),
+	}} {
```
Should also add quantities that is not in this format, CPU can also be specified like `1`, `2`, `0.1`, and would be good to compare those results as well

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_utils_test.go:
```diff
@@ -400,3 +401,202 @@ func TestUtilNormalizedProviderID(t *testing.T) {
+		description:      "valid quantity with unit type",
+		annotations:      map[string]string{memoryKey: "456Mi"},
+		expectedError:    false,
+		expectedQuantity: resource.MustParse("456Mi"),
```
Might also be worth trying some other suffixes in the tests, as `8Gi` is a valid quantity for instance, you can read more about valid formats https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory

elmiko commented:
@JoelSpeed ptal, i think i've covered all the fields we are manually copying through the converters, **even the timestamps!**

after looking deeper, we are not actually copying more from the underlying machines, etc... so i feel good that this test should be covering everything we have added.

i have squashed my extra commit, and i just need to clean up the last few comments about the varying unit types for the memory/cpu stuff.

JoelSpeed commented:
I've reviewed this and am happy for this to merge on the 2 provisos:
- We follow up on the cloud providers machineset controllers to set the memory units properly
- We follow up to do fuzzing and improve the conversion stuff

> @JoelSpeed i also added a few more unit type tests, i'm curious how many of these should we add? i'm avoiding creating some matrix of every possible combo... ;)

I think what you've got is good for now, though we aren't currently testing all of the code in that file, we aren't testing the structured to unstructured converters at the moment so that's why my second proviso is there, this is the kind of stuff that is easily gonna trip someone up as explained offline yesterday

/lgtm

enxebre commented:
related to @JoelSpeed comment above https://github.com/kubernetes/autoscaler/issues/3011
/approve
/retest
/hold
Please un hold once it's verified this passes the validation here https://github.com/openshift/cluster-api-actuator-pkg/pull/140

elmiko commented:
> Please un hold once it's verified this passes the validation here [openshift/cluster-api-actuator-pkg#140](https://github.com/openshift/cluster-api-actuator-pkg/pull/140)

ack, i will spend some time with those tests and then post results here when it's working.

elmiko commented:
i am fairly confident that this code is passing the end-to-end tests from [cluster-api-actuator-pkg#140](https://github.com/openshift/cluster-api-actuator-pkg/pull/140).

i have run the "from/to zero" tests and appear to be working as expected:

```
$ hack/ci-integration.sh -v -focus="zero"  
...
[3] • [SLOW TEST:398.530 seconds]
[3] [Feature:Machines] Autoscaler should
[3] /home/mike/workspace/go/src/github.com/openshift/cluster-api-actuator-pkg/pkg/autoscaler/autoscaler.go:210
[3]   It scales from/to zero
[3]   /home/mike/workspace/go/src/github.com/openshift/cluster-api-actuator-pkg/pkg/autoscaler/autoscaler.go:453
[3] ------------------------------
[3] 
[3] Ran 1 of 5 Specs in 398.534 seconds
[3] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 4 Skipped
[3] PASS

Ginkgo ran 1 suite in 6m43.904044806s
Test Suite Passed
```

i am following up with the full suite but i am seeing non-related issues.

elmiko commented:
/retest

enxebre commented:
/retest

elmiko commented:
as per our conversation on slack, i am going to remove the hold given that the proposed tests are passing.

/hold cancel

elmiko commented:
/retest

elmiko commented:
/test e2e-azure-operator
